python-packages/order_utils/src/zero_ex/order_utils/signature_utils.py

```diff
@@ -0,0 +1,86 @@
+"""Signature utilities."""
+
+from typing import Dict, Tuple
+import json
+from pkg_resources import resource_string
+
+from eth_utils import is_address, to_checksum_address
+from web3 import Web3
+import web3.exceptions
+from web3.utils import datatypes
+
+from zero_ex.dev_utils.type_assertions import assert_is_hex_string
+
+
+# prefer `black` formatting. pylint: disable=C0330
+EXCHANGE_ABI = json.loads(
+    resource_string("zero_ex.contract_artifacts", "artifacts/Exchange.json")
+)["compilerOutput"]["abi"]
+
+network_to_exchange_addr: Dict[str, str] = {
+    "1": "0x4f833a24e1f95d70f028921e27040ca56e09ab0b",
+    "3": "0x4530c0483a1633c7a1c97d2c53721caff2caaaaf",
+    "42": "0x35dd2932454449b14cee11a94d3674a936d5d7b2",
+    "50": "0x48bacb9266a570d521063ef5dd96e61686dbe788",
+}
+
+
+# prefer `black` formatting. pylint: disable=C0330
+def is_valid_signature(
+    provider: Web3.HTTPProvider, data: str, signature: str, signer_address: str
+) -> Tuple[bool, str]:
+    # docstring considered all one line by pylint: disable=line-too-long
+    """Check the validity of the supplied signature.
+
+    Check if the supplied ``signature`` corresponds to signing ``data`` with
+    the private key corresponding to ``signer_address``.
+
+    :param provider: A Web3 provider able to access the 0x Exchange contract.
+    :param data: The hex encoded data signed by the supplied signature.
+    :param signature: The hex encoded signature.
+    :param signer_address: The hex encoded address that signed the data to
+        produce the supplied signature.
+    :rtype: Boolean indicating whether the given signature is valid.
+
+    >>> is_valid_signature(
+    ...     Web3.HTTPProvider("http://127.0.0.1:8545"),
+    ...     '0x6927e990021d23b1eb7b8789f6a6feaf98fe104bb0cf8259421b79f9a34222b0',
+    ...     '0x1B61a3ed31b43c8780e905a260a35faefcc527be7516aa11c0256729b5b351bc3340349190569279751135161d22529dc25add4f6069af05be04cacbda2ace225403',
+    ...     '0x5409ed021d9299bf6814279a6a1411a7e866a631',
+    ... )
+    (True, '')
+    """  # noqa: E501 (line too long)
+    if not isinstance(provider, Web3.HTTPProvider):
```

Comment:
Hm... is there a way to make this more flexible? Literally we should accept any class instance with a public `def make_request(self, method, params)` method used to send JSON RPC calls. Perhaps we can define an abstract class for this?

Reply:
I love the idea but I'm not sure it's that simple.  I removed this type check, and coded up a test of what you're describing, and the test fails.  [Details here](https://gist.github.com/feuGeneA/eed20d7e62bebb93957adc8335006df3).  Maybe the method to implement is actually `request_func()`?  What made you suggest `make_request`?  If it's `request_func()` I'd like to have a test for it (like this one) but I'd have to figure out what's up with that `middleware_stack` argument and how to construct one in the test.  For more context, that `web3/manager.py:89` cited by my gist is [here](https://github.com/ethereum/web3.py/blob/4c52f0d592571077ad2573cefe0587a336685d44/web3/manager.py#L89).

Reply:
I read through Web3.py source and saw that method in both the HTTP and IPC providers.

Reply:
For now, I'm ok with adding a `TODO` comment here for  adding an assertion and shipping this.